### DISCUSSION TITLE
datatree expand/collapse/toggle all children with button

### DISCUSSION THREAD

**Discussion by billcla** [anvilistas/tabulator#155] [Q&A] [answered]
I have implmented the datatree layout which is fantastic for my use case but i'm having issues with downloads, being that unless the children are expanding they dont show in the csv download (which is to be expected based on 6.2 docs).
What I've been trying to do is come up with a way to expand all the children either with a dedicated button or through code at the time the download button is pressed. i have found a few examples where you can either iterate through all the rows and expand the children (but I cant get this to work as for some reason or another i cant identify those rows with children and if there isnt one when you call the toggle() option it fails with a nonetype error), alternatively  there is apparently a way to trigger table options from a function in JS (https://jsfiddle.net/aklaver/74190bvf/8/) but again my inexperience with how JS is translated to python is getting in my way...

Is anyone using datatrees and been able to implment something similar that they would be willing to share please?

My issue is that although i have all children expanded upon loading, if a user collapses one and then downloads the table they cant tell from the csv file that the row should have a child/children and could lead to issues with interpretting the file.

**Comment by billcla** [comment]
As an example of what I've tried, I thought maybe on the event the collaspe button is clicked I could just use the event to expand the treee again to stop users collapsing the rows so they always show in the downloads, as the `all` option of the downloads module doesnt work on datatree children.

`self.tabulator_1.on("dataTreeRowCollapsed", self.row_expand)`

But the row in the function returns an int, being the top level of the tree I presume:

`def row_expand(self, e, row):
    print(f"row: {row}")


returns
`row: 0`

If I use something like:

```
def row_expand(self, e, row):
    print(f"row: {row}")
    row.treeExpand()
```

I get the following including the error:
```
row: 0
AttributeError: 'int' object has no attribute 'treeExpand'
```

What I cant work out is how to retrieve the row like you can with events which are available within the design view such as 'row_click' event which returns the actual row rather than an int, which is what I believe I need to be able to action the `row.treeExpand()` action.

Has anyone had any luck with this?

**Reply by s-cork** [reply]
Is that the right call signature?

Shouldn't the arguments be `row` and `level`.

**Reply by billcla** [reply]
yes, it is, which is whats confusing me more, but when I use the `level` argument i get an `undeclared` warning, but now i'm thinking there might be something else in my code causing that and if `row` in my code above is returning the `int` for the level then maybe the `e` is infact my row... let me check something else!

**Reply by s-cork** [reply]
Yeah there should be no `e` argument

**Reply by billcla** [reply, marked as answer]
It was me... funnily enough! I had a seperate function from a previous test that was causing issues.

Essentially this disables the ability for a user to collaspe a tree by setting the `"dataTreeStartExpanded": True` within the table options, and then `on` the event of someone collapsing the tree:

`self.tabulator_1.on("dataTreeRowCollapsed", self.row_expand)`

Which calls:

```
def row_expand(self, row, level):
    print(f"row: {row}")
    print(f"level: {level}")
    row.treeExpand()
```
This instantly expands the tree to keep it open, which is what I needed in my use case. The `print` lines can be ommited or hashed out as are just there for testing/debugging.

Thanks for listening to an idiot rant... ;-)
